src/box/lua/config/utils/file.lua

```diff
@@ -55,6 +56,117 @@ local function universal_read(file_name, file_kind)
     return data
 end
 
+
+local function get_file_metadata(file_name)
+    local function escape_regex(str)
+        return str:gsub('([%.%+%-%*%?%[%]%(%)%$%^%{%}%|%<%>%~])', '%%%1')
+    end
+
+    -- Tags are only supported for Lua files.
+    if type(file_name) ~= 'string' or not file_name:endswith('.lua') then
+        error(('Invalid file name %q'):format(file_name), 0)
+    end
+
+    local file, err = io.open(file_name, "r")
+    if file == nil then
+        error(('Unable to open %q: %s'):format(file_name, err), 0)
+    end
+
+    local metadata = {}
+
+    local line_count = 0
+    local yaml_string = ''
+    local comment_tag = nil
+    -- Currently the only supported version is 1 and its value is not used.
+    -- luacheck: ignore 311 value assigned to a local variable is unused.
+    local metadata_version = nil
+    local in_comment, in_yaml = false, false
+    while true do
+        local line = file:read("*line")
+        if line == nil then
+            break
+        end
+        line_count = line_count + 1
+
+        -- Skip shebang.
+        if line_count == 1 and line:startswith('#!') then
+            goto next_line
+        end
+
+        -- Skip empty lines.
+        line = line:lstrip()
+        if line == '' then
+            goto next_line
+        end
+
+        if not in_comment and line:startswith('--') then
+            in_comment = true
+        end
+
+        if in_comment then
+            if not line:startswith('--') then
+                break
+            end
+        end
+
+        if in_comment and not in_yaml then
+            comment_tag, metadata_version = line:match(
+                '^(%-%-[%s]+)%-%-%- #tarantool.metadata.v([%d]+)[%s]*$')
+            metadata_version = tonumber(metadata_version)
+
+            if comment_tag ~= nil and metadata_version ~= nil then
+                comment_tag = escape_regex(comment_tag)
+                in_yaml = true
+            end
+        end
+
+        if in_yaml then
+            line = line:match(('^%s(.*)'):format(comment_tag))
+
+            if line == nil then
+                break
+            end
+
+            yaml_string = yaml_string .. '\n' .. line
+
+            if line:startswith('...') then
+                break
+            end
+        end
+
+        ::next_line::
+    end
+    file:close()
+
+    if yaml_string ~= '' then
```

Comment:
If we're `in_yaml == true` here, it means that there is no document end marker. Maybe it worth to handle such a case. (If we do, we should set `in_yaml = false` when meet `...`.)

Reply:
I'm not sure we need to enforce that, because it looks like it is optional, and yaml parser has no problems, when the document has document start tag `---` and no document end tag `...`. I even have a testcase where this is tested.

Reply:
Hmm. Feels like unbalanced parentheses, but OK.